crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+    /// use tokio::signal::unix::SignalKind;
+    ///
+    /// # tokio_test::block_on(async {
+    /// let handle = JoinHandle::new(
+    ///     tokio::spawn(async move {
+    ///         tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    ///     }),
+    ///     SignalKind::terminate(),
+    /// );
+    /// # });
+    /// ```
+    pub fn new(task: tokioJoinHandle<()>, signal: SignalKind) -> Self {
+        JoinHandle {
+            task: OnceCell::from(task),
+            allowed_signal: signal,
+            ..Default::default()
+        }
+    }
+
+    /// This function will block until the task has exited, returning a success or an error
+    async fn wait(&mut self) -> anyhow::Result<()> {
+        return match self.task.take() {
+            Some(handle) => {
+                let result = handle.await;
+                if result.is_err() {
+                    let err = result.unwrap_err();
+                    if err.is_cancelled() {
+                        return anyhow::Ok(());
+                    }
+                    return Err(err.into());
+                }
+                anyhow::Ok(())
+            }
+            None => anyhow::Ok(()),
+        };
+    }
+
+    /// This function will signal the running task with the specified signal
+    /// This is a non-blocking operation
+    fn signal(&self, signal: SignalKind) -> anyhow::Result<()> {
+        if signal != self.allowed_signal && signal != SignalKind::terminate() {
+            return Err(anyhow::anyhow!(
+                "unacceptable signal '{:?}'",
+                signal.as_raw_value()
+            ));
+        }
+        if let Some(handle) = self.task.get() {
+            handle.abort();
+        }
+        return anyhow::Ok(());
+    }
+
+    async fn shutdown(&mut self) {
+        let _ = self.signal(SignalKind::terminate());
+        let _ = self.wait().await;
+    }
+}
+
+impl Future for JoinHandle {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        return match self.task.get() {
+            Some(handle) => {
+                if handle.is_finished() {
+                    Poll::Ready(())
+                } else {
+                    let waker = cx.waker().clone();
+                    tokio::spawn(async move {
+                        tokio::time::sleep(std::time::Duration::from_secs_f32(0.3)).await;
+                        waker.wake();
+                    });
+                    Poll::Pending
+                }
+            }
+            None => Poll::Ready(()),
+        };
+    }
+}
+
+// On drop, send shutdown signal and wait for the task to terminate
+impl Drop for JoinHandle {
+    fn drop(&mut self) {
+        if !self.dropped {
+            let mut this = JoinHandle::default();
+            std::mem::swap(&mut this, self);
+            this.dropped = true;
+            tokio::spawn(async move { this.async_drop().await });
+        }
+    }
+}
```

Comment:
I assume this is rather similar to the `AsyncDropper` wrapper from `async_dropper_simple`. I'm more leaning towards using the `derive` feature from `async_drop` instead of the `simple` version in which case we could use `#[derive(AsyncDrop)]`. What do you think? cc @ryutamago @zcabter

Reply:
Yeah initially I was leaning towards `async_dropper_simple` since it removed the Default requirement as it gets more difficult to define default for large structs and there is a tendency to simply derive non-meaningful defaults makes the code harder to use but it adds invasive boilerplate.

If we use derive, our structs are rather simple (✅ ) and so we just have to be careful to define a meaningful defaults where we can. tldr; derive is fine with me, define meaningful defaults where possible.

Reply:
Alternatively, we don't need to implement drop if the tasks don't panic. Instead, have the jstzd orchestrator registers shutdown/clean up hooks for each task it registers. The orchestrator could control shutdown of tasks based on the state of other tasks. If orchestrator itself panics, it calls shutdown in its async dropper which isolates the async dropper boilerplate. This makes the shutdowns more explicit rather than implicitly cascading, its easier to test and easier to read. wdyt? 

@johnyob